integration_tests/src/main/python/data_gen.py

```diff
@@ -37,15 +39,23 @@ class DataGen:
     """Base class for data generation"""
 
     def __repr__(self):
-        if not self.nullable:
-            return self.__class__.__name__[:-3] + '(not_null)'
-        return self.__class__.__name__[:-3]
+        # repr of DataGens and their children will be used to generate the cache key
+        # make sure it is unique for different DataGens
+        notnull = '(not_null)' if not self.nullable else ''
+        datatype = str(self.data_type)
+        specialcases = ''
+        for (weight, case) in self._special_cases:
+            if (callable(case)):
+                case = case.__code__.co_code
+            specialcases += str(case) + ' ' + str(weight) + ', '
+        specialcases = hashlib.blake2b(specialcases.encode('utf-8'), digest_size=8).hexdigest()
```

Comment:
This shows up in the name of the tests in many cases. I don't think I like having the special cases with a hash in it. I would much rather have a `_cache_repr` with all of the details that is separate from the normal __repr__, and then us the cache version for equality.